plutus-ir/src/Language/PlutusIR/Compiler/Provenance.hs

```diff
@@ -25,8 +25,19 @@ data Provenance a = Original a
                   | TypeBinding String (Provenance a)
                   | DatatypeComponent DatatypeComponent (Provenance a)
                   | NoProvenance
+                  | MultipleSources [Provenance a]
                   deriving (Show, Eq)
 
+-- Needed for LetFloat transformation to merge annotations
+instance Semigroup (Provenance a) where
+  MultipleSources ps1 <> MultipleSources ps2 = MultipleSources (ps1++ps2)
+  x <> MultipleSources ps2 = MultipleSources (x:ps2)
+  MultipleSources ps1 <> x = MultipleSources (ps1++[x])
+  x <> y = MultipleSources [x,y]
+-- Needed for LetFloat transformation to merge annotations
+instance Monoid (Provenance a) where
+  mempty = NoProvenance
```

Comment:
> This is wrong with the semigroup definition you wrote. 

That really depends on what your `==` actually is. It's common for libraries to define `==` over a quotient of a type rather than the type itself, for example see internals of [`pipes`](http://hackage.haskell.org/package/pipes-4.3.13/docs/Pipes-Internal.html#v:observe). So I don't think it's a big deal to break a law provided there is benefit to it, but I think here it's not the case as storing a bunch of `NoProvenance` in `MultipleSources` is noisy and useless.

Can we keep `Provenance` the way it was designed previously, but remove `NoProvenance` from it and just use `[Provenance a]` during compilation? That gives you a natural monoid and no funny tricks are required.

Reply:
That doesn't work because you may want to create a provenance based on the previous provenance. That doesn't work if you have a list of them, unless you have a constructor to wrap them up in... which is what this is.

Reply:
> This is wrong with the semigroup definition you wrote. You need the other one for it to be lawful. Remember we need `x <> mempty == mempty`.

Is this really the monoidal law? I thought x<>mempty ==x

Reply:
Uh yes, sorry, I can't type :sweat_smile: